src/org/recompile/mobile/WavImaAdpcmDecoder.java

```diff
@@ -0,0 +1,370 @@
+/*
+	This file is part of FreeJ2ME.
+
+	FreeJ2ME is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	FreeJ2ME is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with FreeJ2ME.  If not, see http://www.gnu.org/licenses/
+*/
+package org.recompile.mobile;
+
+import java.io.InputStream;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Arrays;
+
+public class WavImaAdpcmDecoder
+{
+	/* Information about this audio format: https://wiki.multimedia.cx/index.php/IMA_ADPCM */
+
+	/* Variables to hold the previously decoded sample and step used, per channel (if needed) */
+	private static int[] prevSample;
+	private static int[] prevStep;
+
+	/* 
+	 * This one acts as the simplest low-pass filter ever, dividing each decoded sample by
+	 * the given amount. Anything from '2' to '4' vastly improves the resulting audio quality 
+	 * by cutting off more of the high range which crackles a lot. '3' is a good middle-ground
+	 * between 2's clearer but still prone to crackling and 4's more consistent but muffled sound.
+	 */
+	private static final int smoothing = 3;
+	
+	private static final int[] ima_step_index_table = 
+	{
+		-1, -1, -1, -1, 2, 4, 6, 8, 
+		-1, -1, -1, -1, 2, 4, 6, 8
+	};
+	
+	private static final int[] ima_step_size_table = 
+	{
+		7, 8, 9, 10, 11, 12, 13, 14, 16, 17,
+		19, 21, 23, 25, 28, 31, 34, 37, 41, 45,
+		50, 55, 60, 66, 73, 80, 88, 97, 107, 118,
+		130, 143, 157, 173, 190, 209, 230, 253, 279, 307,
+		337, 371, 408, 449, 494, 544, 598, 658, 724, 796,
+		876, 963, 1060, 1166, 1282, 1411, 1552, 1707, 1878, 2066,
+		2272, 2499, 2749, 3024, 3327, 3660, 4026, 4428, 4871, 5358,
+		5894, 6484, 7132, 7845, 8630, 9493, 10442, 11487, 12635, 13899,
+		15289, 16818, 18500, 20350, 22385, 24623, 27086, 29794, 32767
+	};
+
+	/* 
+	 * This method will decode IMA WAV ADPCM into linear PCM_S16LE. 
+	 * Note: Largely based on ffmpeg's implementation.
+	 */
+	public static byte[] decodeADPCM(byte[] input, int inputSize, int numChannels, int block_size)
+	{
+		byte[] output;
+		byte adpcmSample;
+		int inputIndex = 0;
+		int outputIndex = 44; /* Give some space for the header */
+		int outputSize, decodedSample; 
+		short curChannel;
+
+		prevSample = new int[2];
+		prevStep = new int[2];
+		
+		outputSize = 44 + (inputSize * 4);
+		output = new byte[outputSize];
+
+		while (inputSize > 0) 
+		{
+			if (inputSize % block_size == 0) 
+			{
+				prevSample[0] = (int) (((input[inputIndex]& 0xFF)) | ((input[inputIndex+1]) << 8) & 0xFF00);
```

Comment:
I knew i was missing something... not only is that little-endian read wrong as you pointed out, there's also quite a few mistakes early on (like not checking this value on the table as you also pointed out, not clamping the value of prevStep in case it is out of bounds, and forcefully passing the wrong data from the header to serve as BlockAlign since dataSize was the only one that resulted in any kind of playback...). BlockAlign/FrameSize/BlockSize was rather confusing due to different sources and documents relating to them with those different names despite all of them being the same thing... i'll refer to it as frameSize from now on

As for what this function is supposed to do... it's meant to be able to decode the whole ADPCM stream while still checking for framesize chunks without depending on a helper function, which is that "% block_size == 0" case on line 80. This is what toggles the function between reading the preamble (because when that 'if' becomes true it means we're at the very beginning of a framesize chunk and the preamble is right in line to be read) and decoding stereo or mono adpcm samples, which also have their own share of issues now that i gave this implementation another look after a while.

The way i have it currently set up, it only needs a few key parts from the header to work, those being number of channels and framesize. Those should theoretically be enough to decode an ADPCM stream, and after making some fixes, that does seem to be the case, as the audio is pretty much 'noiseless' (as in, sounds like ) except for the very beginning of the stream which still has a lot of noise... but more on that later.

About the skip on byte 4 (or byte 3, going from 0 to 3) of the preamble: Apparently it is a reserved byte and usually set to '0' according to [MultimediaWiki](https://wiki.multimedia.cx/index.php/Microsoft_IMA_ADPCM). This would explain why some C implementations i've seen check if this byte is null or not as a means of ascertaining end of stream, as null evaluates to 0 there, it really does seem to serve the purpose of padding.

On the topic of stereo: True, i could use an if statement to cycle through more than two channels, but on a second though we probably won't need to go over 2 channels, as i am yet to find a j2me app that uses stereo adpcm, so i imagine anything beyond that is pretty much non-existent on the platform.


And now back to the noise i mentioned above: It really only happens on the very start of each decoded stream, and is also perfectly reproducible every time. After some back and forth i thought of 2 reasons as to why that might be happening:

> 1 - The predictor and step are set incorrectly on the first chunk to be decoded, although there aren't many documents that delve into this part of the decode process, the only one i could find is again from [MultimediaWiki](https://wiki.multimedia.cx/index.php/IMA_ADPCM) which says those are usually set as '0' at the start of the process, so in the case of Step, it's actually '7' since that's the first index on the table... and as far as i know this is already being done in the code.
>
> 2 (and the most likely reason) - I'm messing something up when passing the stream to be decoded and it's decoding data that shouldn't be decoded (as you pointed out yet again). This one is also helped by the fact that sometimes, only when compiled and ran with Java 8, FreeJ2ME freezes when decoding an ADPCM stream, Java 9 and upwards is rock-solid, but still... there's definitely something wrong in there or else Java 8 would be solid as well.

In any case, there's a commit incoming with many fixes and lots of comments to help understand what's being done in there.